sources/healthdog/src/main.rs

```diff
@@ -0,0 +1,138 @@
+/*!
+# Introduction
+
+Healthdog sends anonymous information about the health of a Bottlerocket host.
+It does so by sending key-value pairs as query params in an HTTP GET request.
+
+Healthdog also has the ability to check that a list of critical services is running.
+It does so using `systemctl` and reports services that are not healthy.
+
+# What it Sends
+
+### The standard set of metrics:
+
+* `sender`: the application sending the report.
+* `event`: the event that invoked the report.
+* `version`: the Bottlerocket version.
+* `variant`: the Bottlerocket variant.
+* `arch`: the machine architecture, e.g.'x86_64' or 'arm'.
+* `region`: the region the machine is running in.
+* `seed`: the seed value used to roll-out updates.
+* `version-lock`: the optional setting that locks Bottlerocket to a certain version.
+* `ignore-waves`: an update setting that allows hosts to update before their seed is reached.
+
+### Additionally, when `healthdog` sends a 'health ping', it adds:
+
+* `is-healthy`: true or false based on whether critical services are running.
+* `failed_services`: a list of critical services that have failed, if any.
+
+# Configuration
+
+Configuration is read from a TOML file, which is generated from Bottlerocket settings:
+
+```toml
+# the url to which healthdog will send metrics information
+metrics_url = "https://example.com/metrics"
+# whether or not healthdog will send metrics. opt-out by setting this to false
+send_metrics = true
+# a list of systemd service names that will be checked
+services = ["apiserver", "containerd", "kubelet"]
+# the region
+aws_region = "us-west-2"
+# the update wave seed
+seed = 1234
+# what version bottlerocket should stay on
+version_lock = "latest"
+# whether bottlerocket should ignore update roll-out timing
+ignore_waves = false
+```
+*/
+
+#![deny(rust_2018_idioms)]
+
+mod args;
+mod config;
+mod error;
+mod healthdog;
+#[cfg(test)]
+mod healthdog_test;
+#[cfg(test)]
+mod main_test;
+mod service_check;
+
+use crate::args::{Command, USAGE};
+use crate::config::Config;
+use crate::error::{Error, Result};
+use crate::healthdog::Healthdog;
+use crate::service_check::{ServiceCheck, SystemdCheck};
+use args::parse_args;
+use bottlerocket_release::BottlerocketRelease;
+use env_logger::Builder;
+use log::{error, trace};
+use snafu::ResultExt;
+use std::sync::Once;
+use std::{env, process};
+
+fn main() -> ! {
+    process::exit(match main_inner(env::args(), Box::new(SystemdCheck {})) {
```

Comment:
```suggestion
    process::exit(match main_inner(env::args(), Box::new(SystemdCheck)) {
```
Don't need empty `{}` for structs with no fields.

Reply:
Seemingly not true?

```
81 |     process::exit(match main_inner(args, Box::new(SystemdCheck)) {
   |                                                   ^^^^^^^^^^^^ did you mean `SystemdCheck { /* fields */ }`?
```